Highlight invalid input in the trading view

ebma opened the pull request:
- [x] Add `formik` dependency
- [x] Add `form` element
- [x] Move validation and submit code to initialisation of `useFormik` hook
- [x] Make the 'Place order' button always clickable
- [x] Add a `selectOnFocus` prop to the `<TradingPrice>` component and enable it in the `TradingForm` so that the content of the "Limit Price"-textfield is selected once focused which makes it easier to edit

Closes #875.

andywer commented on src/components/Trading/TradingForm.tsx:
```diff
@@ -53,22 +58,88 @@ interface Props {
+      manualPrice: undefined,
+      priceMode: "secondary"
+    },
+    validate(values) {
```
We should not inline this validation function, but rather declare it before the component. Should be easy and reduces the component's size.

andywer replied:
👆
This comment still needs to be addressed 😉

ebma replied:
It's actually not that easy because the validation function uses other variables like the `secondaryAmount` which is calculated inside the `TradingForm` component and the calculation requires things like a `tradePair` which is accessed with a hook. So we cannot calculate the `secondaryAmount` in the validation function if it's outside the functional component since we can't get the tradepair. 

Do you have any ideas on how to circumvent this? Maybe I am just missing something.

andywer replied:
Hmm, true. One solution could be currying:

```ts
const prepareValidateFormFields = (tradePair, secondaryAmount) => (values) => {
  // ...
}

// ...

const formik = useFormik<TradingFormValues>({
  // ...
  validate: prepareValidateFormFields(tradePair, secondaryAmount)
})
```

Another solution would be to move the validation method into a hook of its own, so it can use other hooks or we pass those secondary variables when we call the hook (`const validate = useTradingFormValidation(tradePair, secondaryAmount)`).

I think I like the hook-based approach better, but we can talk about it tomorrow.

andywer commented on src/components/Trading/TradingForm.tsx:
```diff
@@ -53,22 +58,88 @@ interface Props {
+
+      return errors
+    },
+    async onSubmit() {
```
Quick idea (and maybe not scope of this PR): How about

```ts
async onSubmit(values: TradingFormValues) {
  try {
    const tx = props.primaryAction === "buy"
      ? await createBuyTx(values, horizon, props.account, props.accountData)
      : await createSellTx(values, horizon, props.account, props.accountData)
    props.sendTransaction(tx)
  } catch (error) {
    trackError(error)
  }
}
```
?

andywer commented:
If we could only find a nice way to use `<Formik>` and its context… We would lose a bit of static typing, since we could not carry the generic types down through the context, BUT we could stop setting `onBlur`, `onChange` & `value` props manually and just wrap our form components in formik helpers that create these props based on the current form context.

Problem right now is that we want to be able to access the current form values in the `TradingForm` component, so we need the `useFormik()` hook and would then need to pour that current formik state into a formik context – we would need a `<FormikProvider>` component or similar that we can render in `TradingForm`.

(Probably not actionable right now, just sharing my thoughts)

andywer commented:
@ebma Please check out this PR. It needs a little more love ♥

ebma commented:
Rebased to latest master and added some changes.

I moved the validation function outside of the component but I'm not quite sure if how I did it makes it better or worse. 
I decided to use the currying method because we can't really use the hooks in this scenario because we need the `values` to initialise them and we only receive these in the actual function that is returned by the hook which on the other hand cannot use the hooks as this would violate the react hook rules.

What I'm trying to say is:
```
const useTradingFormValidation (props, tradePair, secondaryAmount) => 
// here we could use hooks
return (values) => {
// but only here we have the values to correctly initialise them
// which would cause an invariant hook violation
}
```

I still feel like I didn't figure out the best solution but I couldn't think of something better than how it is now.

andywer commented:
@ebma What hooks do we need for form validation that require `values` for initialization? And wouldn’t we then already have this issue now if two hooks were dependent on each other… 😉

ebma commented:
I moved the validation and calculation logic out of the `TradingForm` component.
Still no perfect solution but it works fine with `validateOnChange` set to `false` so let's not waste any more time on this and move on.

andywer commented:
We will use react-hook-form instead of Formik.
